private/buf/bufsync/cached_handler.go

```diff
@@ -0,0 +1,212 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufsync
+
+import (
+	"context"
+
+	"github.com/bufbuild/buf/private/bufpkg/bufmodule/bufmoduleref"
+	registryv1alpha1 "github.com/bufbuild/buf/private/gen/proto/go/buf/alpha/registry/v1alpha1"
+	"github.com/bufbuild/buf/private/pkg/git"
+)
+
+type isBranchSyncedCacheKey struct {
+	moduleIdentityString string
+	branchName           string
+}
+
+type isGitCommitSyncedCacheKey struct {
+	moduleIdentityString string
+	branchName           string
+	gitHash              string
+}
+
+type isProtectedBranchCacheKey struct {
+	moduleIdentityString string
+	branchName           string
+}
+
+type cachedHandler struct {
+	delegate Handler
+
+	isBranchSyncedCache    map[isBranchSyncedCacheKey]struct{}
+	isGitCommitSynedCache  map[isGitCommitSyncedCacheKey]struct{}
+	isProtectedBranchCache map[isProtectedBranchCacheKey]bool
+	isReleaseBranchCache   map[string]bool
+}
+
+func newCachedHandler(delegate Handler) *cachedHandler {
+	return &cachedHandler{
+		delegate:               delegate,
+		isBranchSyncedCache:    make(map[isBranchSyncedCacheKey]struct{}),
+		isGitCommitSynedCache:  make(map[isGitCommitSyncedCacheKey]struct{}),
+		isProtectedBranchCache: make(map[isProtectedBranchCacheKey]bool),
+	}
```

Comment:
You're missing initializing `isReleaseBranchCache`

Reply:
Done